File: packages/python/scripts/parity_snapshot.py
from __future__ import annotations

import json
import os
import shutil
import stat
import sys
import tempfile
from pathlib import Path
from typing import Any

from harness_detect import (
    CheckOptions,
    HarnessCheckResult,
    HarnessSupport,
    HarnessSupportArea,
    HarnessSupportPath,
    HarnessSupportRecord,
    HarnessSupportScope,
    check_harness,
    detect_harnesses,
    get_harness_support,
    list_harness_support,
)


def main() -> None:
    try:
        run()
    except Exception as error:  # noqa: BLE001 - CLI helper should print concise errors.
        print(f"parity_snapshot: {error}", file=sys.stderr)
        raise SystemExit(1) from error


def run() -> None:
    parsed = json.loads(read_input())
    cases = parsed.get("cases")
    if not isinstance(cases, list):
        raise ValueError("Parity input must be an object with a cases array.")

    roots = create_sandbox()
    try:
        output = {
            "version": parsed.get("version"),
            "cases": [run_case(case, roots) for case in cases],
        }
        print(json.dumps(output, indent=2))
    finally:
        shutil.rmtree(roots["TMP"], ignore_errors=True)


def read_input() -> str:
    if len(sys.argv) > 1 and sys.argv[1] != "-":
        return Path(sys.argv[1]).read_text(encoding="utf-8")

    raw = sys.stdin.read()
    if not raw.strip():
        raise ValueError(
            "Expected JSON parity cases from stdin or a file path argument."
        )
    return raw


def create_sandbox() -> dict[str, str]:
    temp_root = tempfile.mkdtemp(prefix="harness-detect-parity-")
    roots = {
        "TMP": temp_root,
        "HOME": os.path.join(temp_root, "home"),
        "CWD": os.path.join(temp_root, "cwd"),
        "BIN": os.path.join(temp_root, "bin"),
    }
    for path in roots.values():
        os.makedirs(path, exist_ok=True)
    return roots


def run_case(raw_case: dict[str, Any], roots: dict[str, str]) -> dict[str, Any]:
    case = expand_value(raw_case, roots)
    if should_skip_platform(case):
        return {"id": case["id"], "operation": case["operation"], "skipped": True}

    setup = prepare_setup(case.get("setup", []), roots["TMP"])
    apply_setup(setup)
    try:
        options = CheckOptions(env=dict(case.get("env", {})), cwd=case.get("cwd"))
        operation = case["operation"]
        if operation == "checkHarness":
            result = normalize_check_result(
                check_harness(case["input"], options), roots
            )
        elif operation == "detectHarnesses":
            result = normalize_detect_results(detect_harnesses(options), roots)
        elif operation == "getHarnessSupport":
            result = normalize_support_record(get_harness_support(case["input"]))
        elif operation == "listHarnessSupport":
            result = normalize_support_list(list_harness_support())
        else:
            raise ValueError(f"Unsupported parity operation: {operation}")

        return {"id": case["id"], "operation": operation, "result": result}
    finally:
        cleanup_setup(setup)


def expand_value(value: Any, roots: dict[str, str]) -> Any:
    if isinstance(value, str):
        return expand_string(value, roots)
    if isinstance(value, list):
        return [expand_value(item, roots) for item in value]
    if isinstance(value, dict):
        return {key: expand_value(item, roots) for key, item in value.items()}
    return value


def expand_string(value: str, roots: dict[str, str]) -> str:
    for key, root in roots.items():
        value = value.replace("${" + key + "}", root)
    return value


def should_skip_platform(case: dict[str, Any]) -> bool:
    platforms = case.get("platforms")
    return isinstance(platforms, list) and current_platform() not in platforms


def current_platform() -> str:
    if sys.platform == "darwin":
        return "darwin"
    if sys.platform == "win32":
        return "win32"
    return sys.platform


def prepare_setup(
    setup: list[dict[str, Any]], sandbox_root: str
) -> list[dict[str, str]]:
    return [
        {
            "type": entry["type"],
            "path": resolve_sandbox_path(entry["path"], sandbox_root),
            "content": entry.get("content", ""),
        }
        for entry in setup
    ]


def resolve_sandbox_path(target_path: str, sandbox_root: str) -> str:
    absolute = os.path.abspath(target_path)
    sandbox = os.path.abspath(sandbox_root)
    if os.path.commonpath([sandbox, absolute]) != sandbox:
        raise ValueError(f"Parity setup path escapes temp root: {target_path}")
    return absolute


def apply_setup(setup: list[dict[str, str]]) -> None:
    for entry in setup:
        path = entry["path"]
        if entry["type"] == "dir":
            os.makedirs(path, exist_ok=True)
        elif entry["type"] in {"file", "executable"}:
            os.makedirs(os.path.dirname(path), exist_ok=True)
            Path(path).write_text(entry["content"], encoding="utf-8")
            if entry["type"] == "executable" and sys.platform != "win32":
                os.chmod(path, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
        else:
            raise ValueError(f"Unsupported parity setup type: {entry['type']}")


def cleanup_setup(setup: list[dict[str, str]]) -> None:
    for entry in reversed(setup):
        path = entry["path"]
        if not os.path.exists(path):
            continue
        if entry["type"] == "dir":
            shutil.rmtree(path, ignore_errors=True)
        else:
            try:
                os.remove(path)
            except FileNotFoundError:
                pass


def normalize_check_result(
    result: HarnessCheckResult, roots: dict[str, str]
) -> dict[str, Any]:
    return {
        "key": result.key,
        "installed": bool(result.installed),
        "executablePath": normalize_path_value(result.executable_path, roots),
        "matchedPathIds": sorted(path.id for path in result.matched_paths),
        "paths": sorted(
            (
                {
                    "id": path.id,
                    "path": normalize_path_value(path.path, roots),
                    "exists": bool(path.exists),
                }
                for path in result.paths
            ),
            key=lambda path: path["id"],
        ),
    }


def normalize_detect_results(
    results: list[HarnessCheckResult], roots: dict[str, str]
) -> dict[str, Any]:
    normalized_results = sorted(
        (normalize_check_result(result, roots) for result in results),
        key=lambda result: result["key"],
    )
    installed_keys = sorted(
        result["key"] for result in normalized_results if result["installed"]
    )
    return {
        "count": len(results),
        "installedCount": len(installed_keys),
        "installedKeys": installed_keys,
        "results": normalized_results,
    }


def normalize_path_value(value: str | None, roots: dict[str, str]) -> str | None:
    if value is None:
        return None

    normalized = os.path.abspath(value)
    replacements = sorted(
        (("${" + key + "}", os.path.abspath(root)) for key, root in roots.items()),
        key=lambda item: len(item[1]),
        reverse=True,
    )
    for placeholder, root in replacements:
        if normalized == root:
            return placeholder
        prefix = root + os.sep
        if normalized.startswith(prefix):
            return placeholder + normalized[len(root) :]
    return normalized


def normalize_support_path(path: HarnessSupportPath) -> dict[str, Any]:
    return {
        "id": path.id,
        "kind": path.kind,
        "template": path.template,
        "platforms": sorted(path.platforms) if path.platforms else None,
        "description": path.description or None,
    }


def normalize_support_leaf(leaf: HarnessSupportScope) -> dict[str, Any]:
    return {
        "status": leaf.status,
        "confidence": leaf.confidence,
        "notes": leaf.notes or None,
        "sources": sorted(leaf.sources),
        "paths": sorted(
            (normalize_support_path(path) for path in leaf.paths),
            key=lambda path: path["id"],
        ),
    }


def normalize_support_area(area: HarnessSupportArea) -> dict[str, Any]:
    return {
        "global": normalize_support_leaf(area.global_),
        "local": normalize_support_leaf(area.local),
    }


def normalize_support(support: HarnessSupport) -> dict[str, Any]:
    return {
        "config": normalize_support_area(support.config),
        "skills": normalize_support_area(support.skills),
        "commands": normalize_support_area(support.commands),
        "agents": normalize_support_area(support.agents),
        "dotAgents": normalize_support_area(support.dot_agents),
    }


def normalize_support_record(record: HarnessSupportRecord) -> dict[str, Any]:
    return {
        "key": record.key,
        "name": record.name,
        "support": normalize_support(record.support),
    }


def normalize_support_list(records: list[HarnessSupportRecord]) -> dict[str, Any]:
    normalized_records = sorted(
        (normalize_support_record(record) for record in records),
        key=lambda record: record["key"],
    )
    return {"count": len(normalized_records), "records": normalized_records}


if __name__ == "__main__":
    main()

File: packages/typescript/scripts/parity-snapshot.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import { checkHarness, detectHarnesses, getHarnessSupport, listHarnessSupport } from "../src/index.js";

function readInput(sourceArg) {
  if (sourceArg && sourceArg !== "-") {
    const sourcePath = path.resolve(process.cwd(), sourceArg);
    return fs.readFileSync(sourcePath, "utf8");
  }

  return fs.readFileSync(0, "utf8");
}

function parseInput(raw) {
  if (!raw.trim()) {
    throw new Error("Expected JSON parity cases from stdin or a file path argument.");
  }

  const parsed = JSON.parse(raw);
  const cases = Array.isArray(parsed) ? parsed : parsed.cases;

  if (!Array.isArray(cases)) {
    throw new Error("Parity input must be an array of cases or an object with a cases array.");
  }

  return {
    version: Array.isArray(parsed) ? null : parsed.version ?? null,
    cases
  };
}

function createSandbox() {
  const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), "harness-detect-parity-"));
  const roots = {
    TMP: tempRoot,
    HOME: path.join(tempRoot, "home"),
    CWD: path.join(tempRoot, "cwd"),
    BIN: path.join(tempRoot, "bin")
  };

  for (const dirPath of Object.values(roots)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }

  return roots;
}

function expandString(value, roots) {
  return value.replace(/\$\{(TMP|HOME|CWD|BIN)\}/g, (_, key) => roots[key]);
}

function expandValue(value, roots) {
  if (typeof value === "string") {
    return expandString(value, roots);
  }

  if (Array.isArray(value)) {
    return value.map((entry) => expandValue(entry, roots));
  }

  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, entry]) => [key, expandValue(entry, roots)])
    );
  }

  return value;
}

function expandCase(testCase, roots) {
  return {
    id: testCase.id,
    operation: testCase.operation,
    ...(testCase.input === undefined ? {} : { input: expandValue(testCase.input, roots) }),
    ...(testCase.platforms === undefined ? {} : { platforms: expandValue(testCase.platforms, roots) }),
    ...(testCase.cwd === undefined ? {} : { cwd: expandValue(testCase.cwd, roots) }),
    ...(testCase.env === undefined ? {} : { env: expandValue(testCase.env, roots) }),
    ...(testCase.setup === undefined ? {} : { setup: expandValue(testCase.setup, roots) })
  };
}

function normalizePathString(value, roots) {
  if (typeof value !== "string") {
    return value ?? null;
  }

  const replacements = Object.entries(roots)
    .map(([key, rootPath]) => [key, path.resolve(rootPath)])
    .sort((left, right) => right[1].length - left[1].length);

  let normalized = path.resolve(value);
  for (const [key, rootPath] of replacements) {
    if (normalized === rootPath || normalized.startsWith(`${rootPath}${path.sep}`)) {
      normalized = "${" + key + "}" + normalized.slice(rootPath.length);
      break;
    }
  }

  return normalized;
}

function isWithinRoot(rootPath, candidatePath) {
  const relativePath = path.relative(rootPath, candidatePath);
  return relativePath === "" || (!relativePath.startsWith("..") && !path.isAbsolute(relativePath));
}

function resolveSandboxPath(targetPath, sandboxRoot) {
  if (typeof targetPath !== "string" || targetPath.length === 0) {
    throw new Error("Parity setup entries must include a non-empty path.");
  }

  if (typeof sandboxRoot !== "string" || sandboxRoot.length === 0) {
    throw new Error(`Parity setup path requires a tempRoot sandbox: ${targetPath}`);
  }

  const sandboxRealPath = fs.realpathSync.native(sandboxRoot);
  const absoluteTargetPath = path.resolve(targetPath);
  const missingSegments = [];
  let existingPath = absoluteTargetPath;

  while (!fs.existsSync(existingPath)) {
    const parentPath = path.dirname(existingPath);
    if (parentPath === existingPath) {
      throw new Error(`Could not resolve a parity setup parent for ${targetPath}`);
    }

    missingSegments.unshift(path.basename(existingPath));
    existingPath = parentPath;
  }

  const existingStat = fs.lstatSync(existingPath);
  if (existingPath === absoluteTargetPath && existingStat.isSymbolicLink()) {
    throw new Error(`Refusing parity setup symlink target: ${targetPath}`);
  }

  const resolvedExistingPath = fs.realpathSync.native(existingPath);
  const resolvedTargetPath = path.join(resolvedExistingPath, ...missingSegments);
  if (!isWithinRoot(sandboxRealPath, resolvedTargetPath)) {
    throw new Error(`Parity setup path escapes temp root: ${targetPath}`);
  }

  return absoluteTargetPath;
}

function prepareSetup(setup, sandboxRoot) {
  return setup.map((entry) => ({
    ...entry,
    path: resolveSandboxPath(entry.path, sandboxRoot)
  }));
}

function ensureParent(filePath) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
}

function applySetupEntry(entry) {
  if (entry.type === "dir") {
    fs.mkdirSync(entry.path, { recursive: true });
    return;
  }

  if (entry.type === "file" || entry.type === "executable") {
    ensureParent(entry.path);
    fs.writeFileSync(entry.path, entry.content ?? "");

    if (entry.type === "executable" && process.platform !== "win32") {
      fs.chmodSync(entry.path, 0o755);
    }

    return;
  }

  throw new Error(`Unsupported parity setup type: ${entry.type}`);
}

function cleanupSetupEntry(entry) {
  if (!entry?.path || !fs.existsSync(entry.path)) {
    return;
  }

  if (entry.type === "dir") {
    fs.rmSync(entry.path, { recursive: true, force: true });
    return;
  }

  fs.rmSync(entry.path, { force: true });
}

function withSetup(setup, fn) {
  const appliedEntries = [];

  try {
    for (const entry of setup) {
      applySetupEntry(entry);
      appliedEntries.push(entry);
    }

    return fn();
  } finally {
    for (const entry of [...appliedEntries].reverse()) {
      cleanupSetupEntry(entry);
    }
  }
}

function sortById(items) {
  return [...items].sort((left, right) => left.id.localeCompare(right.id));
}

function sortStrings(values) {
  return [...values].sort((left, right) => left.localeCompare(right));
}

function normalizeSupportPath(pathSpec) {
  return {
    id: pathSpec.id,
    kind: pathSpec.kind,
    template: pathSpec.template,
    platforms: Array.isArray(pathSpec.platforms) ? sortStrings(pathSpec.platforms) : null,
    description: pathSpec.description ?? null
  };
}

function normalizeSupportLeaf(leaf) {
  return {
    status: leaf.status,
    confidence: leaf.confidence,
    notes: leaf.notes ?? null,
    sources: sortStrings(leaf.sources),
    paths: [...leaf.paths].map(normalizeSupportPath).sort((left, right) => left.id.localeCompare(right.id))
  };
}

function normalizeSupportArea(area) {
  return {
    global: normalizeSupportLeaf(area.global),
    local: normalizeSupportLeaf(area.local)
  };
}

function normalizeSupport(support) {
  return {
    config: normalizeSupportArea(support.config),
    skills: normalizeSupportArea(support.skills),
    commands: normalizeSupportArea(support.commands),
    agents: normalizeSupportArea(support.agents),
    dotAgents: normalizeSupportArea(support.dotAgents)
  };
}

function normalizeSupportRecord(record) {
  return {
    key: record.key,
    name: record.name,
    support: normalizeSupport(record.support)
  };
}

function normalizeSupportList(records) {
  const normalizedRecords = [...records]
    .map((record) => normalizeSupportRecord(record))
    .sort((left, right) => left.key.localeCompare(right.key));

  return {
    count: normalizedRecords.length,
    records: normalizedRecords
  };
}

function normalizeHarnessResult(result, roots) {
  return {
    key: result.key,
    installed: Boolean(result.installed),
    executablePath: normalizePathString(result.executablePath, roots),
    matchedPathIds: sortStrings(result.matchedPaths.map((entry) => entry.id)),
    paths: sortById(result.paths).map((entry) => ({
      id: entry.id,
      path: normalizePathString(entry.path, roots),
      exists: Boolean(entry.exists)
    }))
  };
}

function normalizeDetectResults(results, roots) {
  const normalizedResults = [...results]
    .map((result) => normalizeHarnessResult(result, roots))
    .sort((left, right) => left.key.localeCompare(right.key));

  const installedKeys = normalizedResults
    .filter((result) => result.installed)
    .map((result) => result.key);

  return {
    count: normalizedResults.length,
    installedCount: installedKeys.length,
    installedKeys,
    results: normalizedResults
  };
}

function runCase(rawCase, roots) {
  const testCase = expandCase(rawCase, roots);

  if (Array.isArray(testCase.platforms) && !testCase.platforms.includes(process.platform)) {
    return {
      id: testCase.id,
      operation: testCase.operation,
      skipped: true
    };
  }

  const options = {
    env: testCase.env ?? {},
    cwd: testCase.cwd
  };

  const setup = prepareSetup(Array.isArray(testCase.setup) ? testCase.setup : [], roots.TMP);

  return withSetup(setup, () => {
    if (testCase.operation === "checkHarness") {
      return {
        id: testCase.id,
        operation: testCase.operation,
        result: normalizeHarnessResult(checkHarness(testCase.input, options), roots)
      };
    }

    if (testCase.operation === "detectHarnesses") {
      return {
        id: testCase.id,
        operation: testCase.operation,
        result: normalizeDetectResults(detectHarnesses(options), roots)
      };
    }

    if (testCase.operation === "getHarnessSupport") {
      return {
        id: testCase.id,
        operation: testCase.operation,
        result: normalizeSupportRecord(getHarnessSupport(testCase.input))
      };
    }

    if (testCase.operation === "listHarnessSupport") {
      return {
        id: testCase.id,
        operation: testCase.operation,
        result: normalizeSupportList(listHarnessSupport())
      };
    }

    throw new Error(`Unsupported parity operation: ${testCase.operation}`);
  });
}

function main() {
  const input = parseInput(readInput(process.argv[2]));
  const roots = createSandbox();

  try {
    const output = {
      version: input.version,
      cases: input.cases.map((testCase) => runCase(testCase, roots))
    };

    process.stdout.write(`${JSON.stringify(output, null, 2)}\n`);
  } finally {
    fs.rmSync(roots.TMP, { recursive: true, force: true });
  }
}

main();
